### PR TITLE
Docs: New - Guidelines, conventions & file's structure ( components. …

### PR DESCRIPTION
**Pull request by iNewLegend**

### CODE REVIEW

File: docs/core/common/assets/js/api/core/commands.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+     // Run's 'example' command from 'custom-component'.
+     result = $e.run( 'custom-component/example', {
+        property: 'value',
+     } );
+     
+     // Output command run result.
+     console.log( 'e-commands-eg-1-result: ', result );
+    ```
+## Guidelines, conventions & file's structure
+  * Each command, owned by a [component](../core/components.md#guidelines-conventions--files-structure).
```
Comment (matipojo) [inline docs/core/common/assets/js/api/core/commands.md:75]:
```suggestion
  * Each command, should be owned by a [component](../core/components.md#guidelines-conventions--files-structure).
```

File: docs/core/common/assets/js/api/core/commands.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+  * Currently, there are _3_ main *base/types*: 
+    * Commands - Base class: `$e.modules.CommandBase` - `[USER]` commands that represent user actions.
+    * Commands internal - Base class:  `$e.modules.CommandInternalBase` - `[INTERNAL]` for internal usage.
+    * Commands data - Base class: `$e.modules.CommandData` - `[DATA]` commands for communicate with the _data\cache\backend_.
+  * Each [component](../core/components.md#guidelines-conventions--files-structure), can extend few methods `defaultCommands`, `defaultCommandsInternal`, `defaultData`
```
Comment (matipojo) [inline docs/core/common/assets/js/api/core/commands.md:80]:
```suggestion
  * Each [component](../core/components.md#guidelines-conventions--files-structure), can override few methods `defaultCommands`, `defaultCommandsInternal`, `defaultData` 
```

File: docs/core/common/assets/js/api/core/commands.md [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    ```
+## Guidelines, conventions & file's structure
+  * Each command, owned by a [component](../core/components.md#guidelines-conventions--files-structure).
+  * Currently, there are _3_ main *base/types*: 
+    * Commands - Base class: `$e.modules.CommandBase` - `[USER]` commands that represent user actions.
+    * Commands internal - Base class:  `$e.modules.CommandInternalBase` - `[INTERNAL]` for internal usage.
+    * Commands data - Base class: `$e.modules.CommandData` - `[DATA]` commands for communicate with the _data\cache\backend_.
+  * Each [component](../core/components.md#guidelines-conventions--files-structure), can extend few methods `defaultCommands`, `defaultCommandsInternal`, `defaultData` 
+  methods which are used to import the command(s), according to their type.
+  * The commands imported via built-in method called `importCommands`.
```
Comment (matipojo) [inline docs/core/common/assets/js/api/core/commands.md:82]:
```suggestion
  * The commands should be imported via built-in method called `importCommands`.
```

File: docs/core/common/assets/js/api/core/commands.md [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        ```html class:"lineNo"
+        1  📦 component
+        2  │   📜 component.js
+        3  │
+        4  └───📂 commands
+        5  │   │   📜 index.js ( has all the commands exported )
+        6  │   │   📜 exmaple-command.js
+        7  │   │   ...
+        ```
+        `index.js` file at line *5*:
```
Comment (matipojo) [inline docs/core/common/assets/js/api/core/commands.md:93]:
```suggestion
        `commands/index.js` file at line *5*:
```

File: docs/core/common/assets/js/api/core/commands.md [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+            ```html class:"lineNo"
+            1  📦 component
+            2  │   📜 component.js
+            3  │
+            4  └───📂 commands-internal
+            5  │   │   📜 index.js ( has all the commands exported )
+            6  │   │   📜 internal-command.js
+            7  │   │   ...
+            ```
+        * `index.js` file at line *5*:
```
Comment (matipojo) [inline docs/core/common/assets/js/api/core/commands.md:125]:
```suggestion
        * `commands-internal/index.js` file at line *5*:
```

File: docs/core/common/assets/js/api/core/commands.md [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+                getNamespace() {
+                    return 'component-name';
+                }
+    
+                defaultData() {
+                    return this.importCommands( commandsData );
+                }
+            }
+            ```
+> ### **Note:** further information about [`{$e.modules.CommandBase}`](../modules/command-base.full.md)**class**.
```
Comment (matipojo) [inline docs/core/common/assets/js/api/core/commands.md:172]:
```suggestion
 ### **Note:** further information about [`{$e.modules.CommandBase}`](../modules/command-base.full.md)**class**.
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: docs/core/common/assets/js/api/core/commands.md (hunk 1 above)
Comment (iNewLegend) [reply docs/core/common/assets/js/api/core/commands.md]: Handled.

File: docs/core/common/assets/js/api/core/commands.md (hunk 2 above)
Comment (iNewLegend) [reply docs/core/common/assets/js/api/core/commands.md]: Handled.

File: docs/core/common/assets/js/api/core/commands.md (hunk 3 above)
Comment (iNewLegend) [reply docs/core/common/assets/js/api/core/commands.md]: Handled.

File: docs/core/common/assets/js/api/core/commands.md (hunk 4 above)
Comment (iNewLegend) [reply docs/core/common/assets/js/api/core/commands.md]: Handled.

File: docs/core/common/assets/js/api/core/commands.md (hunk 5 above)
Comment (iNewLegend) [reply docs/core/common/assets/js/api/core/commands.md]: Handled.

File: docs/core/common/assets/js/api/core/commands.md (hunk 6 above)
Comment (iNewLegend) [reply docs/core/common/assets/js/api/core/commands.md]: Handled.

File: docs/core/common/assets/js/api/core/commands.md [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+            ```html class:"lineNo"
+            1  📦 component
+            2  │   📜 component.js
+            3  │
+            4  └───📂 commands-data
+            5  │   │   📜 index.js ( has all the commands exported )
+            6  │   │   📜 data-command.js
+            7  │   │   ...
+            ```
+        * `index.js` file at line *5*:
```
Comment (matipojo) [inline docs/core/common/assets/js/api/core/commands.md:154]: Add the dir name for all `index.js`. here and next.

**Review by matipojo — CHANGES_REQUESTED** [review]

File: docs/core/common/assets/js/api/core/commands.md (hunk 7 above)
Comment (iNewLegend) [reply docs/core/common/assets/js/api/core/commands.md]: Handled.

File: docs/core/common/assets/js/api/core/components.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    1  📦 document
+    2  │   📜 component.js
+    3  │   📜 index.js           ( has all sub-components exported )
+    4  │
+    5  └───📂 elements
+    6  │   │   📜 component.js
+    7  │   │   |   ...
+    ```
+    
+    `index.js` file at line *3*:
```
Comment (matipojo) [inline docs/core/common/assets/js/api/core/components.md:79]: :(

File: docs/core/common/assets/js/api/core/components.md [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    ```javascript
+    export { default as ElementsComponent } from './elements/component.js';
+    ```    
+    `component.js` file at line *2*:
```
Comment (matipojo) [inline docs/core/common/assets/js/api/core/components.md:83]: :(

File: docs/core/common/assets/js/api/core/hooks.md [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+    22		const { footerSaver } = $e.components.get( 'document/save' );
+    23
+    24		footerSaver.setMenuItems( args.container.document );
+    25
+    26		footerSaver.refreshWpPreview();
+    27	}
+    28 }
+    29
+    30 export default FooterSaverRefreshMenu; // Required - comment should be deleted.
+
```
Comment (matipojo) [inline docs/core/common/assets/js/api/core/hooks.md:178]: End javascript?

**Review by matipojo — CHANGES_REQUESTED** [review]

**Review by matipojo — DISMISSED** [review]

File: docs/core/common/assets/js/api/core/components.md [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+                  };
+              },
+            };
+        }
+     }
+     
+     // Register the new component.
+     $e.components.register( new CustomComponent() );
+     
+     // Run's 'example' command from 'custom-component'.
```
Comment (kobizz) [inline docs/core/common/assets/js/api/core/components.md:54]:
```suggestion
     // Runs 'example' command from 'custom-component'.
```

File: docs/core/common/assets/js/api/core/components.md [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+     result = $e.run( 'custom-component/example', {
+        property: 'value',
+     } );
+     
+     // Output command run result.
+     console.log( 'e-components-eg-1-result: ', result );
+    
+    ```
+## Guidelines, conventions & file's structure
+  * You can view your component as namespace, that holds your [commands](../core/commands.md#guidelines-conventions--files-structure), [hooks](../core/hooks.md#guidelines-conventions--files-structure), [routes](#UPDATE_WHEN_READY), [tabs](#UPDATE_WHEN_READY), [shortcuts](#UPDATE_WHEN_READY) & [utils](#UPDATE_WHEN_READY).
```
Comment (kobizz) [inline docs/core/common/assets/js/api/core/components.md:64]:
```suggestion
  * You can view your component as a namespace that holds your [commands](../core/commands.md#guidelines-conventions--files-structure), [hooks](../core/hooks.md#guidelines-conventions--files-structure), [routes](#UPDATE_WHEN_READY), [tabs](#UPDATE_WHEN_READY), [shortcuts](#UPDATE_WHEN_READY) & [utils](#UPDATE_WHEN_READY).
```

File: docs/core/common/assets/js/api/core/components.md [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+  * Component class file should be named `component.js`
+  * Component folder name should be named as component namespace or sub component namespace.
```
Comment (kobizz) [inline docs/core/common/assets/js/api/core/components.md:66]:
```suggestion
  * Component folder name should be named as a component namespace or a sub-component namespace.
```

Comment (kobizz) [reply docs/core/common/assets/js/api/core/components.md:66]: We have to fix some of our current files structure according to these conventions

File: docs/core/common/assets/js/api/core/hooks.md [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+	| Method                               | Params                                                                                          |  Returns              | Description                                              
+	|--------------------------------------|-------------------------------------------------------------------------------------------------|-----------------------|-----------------------------------------------------------
+	| `$e.hooks.activate()`                |                                                                                                 |                       | Activate all hooks. 
+	| `$e.hooks.deactivate()`              |                                                                                                 |                       | Deactivate all hooks. 
+	| `$e.hooks.getAll()`                  |                                                                                                 | `{Array}`             | Receive all loaded hooks. 
+	| `$e.hooks.register()`                | `{String}` *type*, `{String}` *event*, `{HookBase}` *instance*                                  | `{Object}` *callback* | Register hook.
```
Comment (kobizz) [inline docs/core/common/assets/js/api/core/hooks.md:12]:
```suggestion
	| `$e.hooks.register()`                | `{String}` *type*, `{String}` *event*, `{HookBase}` *instance*                                  | `{Object}` *callback* | Register a hook. 
```

File: docs/core/common/assets/js/api/core/hooks.md [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+	|--------------------------------------|-------------------------------------------------------------------------------------------------|-----------------------|-----------------------------------------------------------
+	| `$e.hooks.activate()`                |                                                                                                 |                       | Activate all hooks. 
+	| `$e.hooks.deactivate()`              |                                                                                                 |                       | Deactivate all hooks. 
+	| `$e.hooks.getAll()`                  |                                                                                                 | `{Array}`             | Receive all loaded hooks. 
+	| `$e.hooks.register()`                | `{String}` *type*, `{String}` *event*, `{HookBase}` *instance*                                  | `{Object}` *callback* | Register hook. 
+	| `$e.hooks.run()`                     | `{String}` *type*, `{String}` *event*, `{String}` *command*, `{Object}` *args*, `{*}` *result*  | `{Boolean}`           | Run's a hook.
```
Comment (kobizz) [inline docs/core/common/assets/js/api/core/hooks.md:13]: > Run's a hook

You can write either "Run" or "Runs" (depending on your meaning), but not "Run's".

File: docs/core/common/assets/js/api/core/hooks.md [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+	| `$e.hooks.registerDataDependency()`  | `{HookBase}` *instance*                                                                         | `{Object}` *callback* | Register data hook that's run before the command as dependency.  
+	| `$e.hooks.registerUIAfter()`         | `{HookBase}` *instance*                                                                         | `{Object}` *callback* | Register UI hook that's run after the commands run.  
+	| `$e.hooks.registerUICatch()`         | `{HookBase}` *instance*                                                                         | `{Object}` *callback* | Register UI hook that's run when the command fails.  
+	| `$e.hooks.registerUIBefore()`        | `{HookBase}` *instance*                                                                         | `{Object}` *callback* | Register UI hook that's run before the command.  
+	| `$e.hooks.runDataAfter()`            | `{String}` *command*, `{Object}` *args*, `{*}` *result*                                         | `{Boolean}`           | Run data hook that's run after the command.
```
Comment (kobizz) [inline docs/core/common/assets/js/api/core/hooks.md:20]:
```suggestion
	| `$e.hooks.runDataAfter()`            | `{String}` *command*, `{Object}` *args*, `{*}` *result*                                         | `{Boolean}`           | Run a data hook that runs after the command.
```

Comment (kobizz) [reply docs/core/common/assets/js/api/core/hooks.md:20]: The same goes for the rest of the lines

File: docs/core/common/assets/js/api/core/hooks.md [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+	| `$e.hooks.runDataDependency()`       | `{String}` *command*, `{Object}` *args*, `{*}` *result*                                         | `{Boolean}`           | Run data hook that's run before the command as dependency.
+	| `$e.hooks.runUIAfter()`              | `{String}` *command*, `{Object}` *args*, `{*}` *result*                                         | `{Boolean}`           | Run UI hook that's run after the commands run.
+	| `$e.hooks.runUICatch()`              | `{String}` *command*, `{Object}` *args*, `{*}` *result*                                         | `{Boolean}`           | Run UI hook that's run when the command fails.
+	| `$e.hooks.runUIBefore()`             | `{String}` *command*, `{Object}` *args*, `{*}` *result*                                         | `{Boolean}`           | Run UI hook that's run before the command.
+
+## Guidelines, conventions & file's structure 
+  * Each hook, owned by a [component](../core/components.md#guidelines-conventions--files-structure).
```
Comment (kobizz) [inline docs/core/common/assets/js/api/core/hooks.md:28]:
```suggestion
  * Each hook is owned by a [component](../core/components.md#guidelines-conventions--files-structure).
```

File: docs/core/common/assets/js/api/core/hooks.md [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+    |`{TYPE}`                | `ui` or `data` depends on the hook  | `ui`
+    |`{COMMAND}`             | which command to hook               | `document/elements/settings`
+    |`{FILE_NAME}`           | kebab case, name is description oh what the hook does | `footer-saver-refresh-menu.js`
+    |`{FILE_NAME_CAMEL_CASE}`| camel case, of `{FILE_NAME}`        | `FooterSaverRefreshMenu`
+    |`{FILE_NAME_WITHOUT_JS}`| `{FILE_NAME}` without `.js`         | `footer-saver-refresh-menu`
+    |`{FILE_PATH}`           | `{TYPE}/{COMMAND}/{FILE_NAME}`      | `ui/document/elements/settings/footer-saver-refresh-menu.js`
+    |`{CONTAINER_TYPE}`      | optional, gain performance if container type is known in advance | `document`
+    > Example
+    ```javascript class:"lineNo"
+    1 // ui/document/elements/settings/footer-saver-refresh-menu.js - This is line should be deleted - just for the exmaple.
```
Comment (kobizz) [inline docs/core/common/assets/js/api/core/hooks.md:148]:
```suggestion
    1 // ui/document/elements/settings/footer-saver-refresh-menu.js - This line should be deleted - just for the example.
```

**Review by kobizz — CHANGES_REQUESTED** [review]

File: docs/core/common/assets/js/api/core/components.md (hunk 13 above)
Comment (iNewLegend) [reply docs/core/common/assets/js/api/core/components.md:66]: After we done with 'conventions' i will add to my self a task: fix new conventions in existing code.

File: docs/core/common/assets/js/api/core/commands.md [hunk 19]
```diff
@@ -0,0 +1,174 @@
...
+
+The full list of commands, including custom & 3rd commands, is available via: `$e.commands.getAll();`
+
+*  **Description**: `$e.commands` API is a manager of `$e.commands`, allow you to create custom **commands** that runs by `$e.run()`, each components have his own unique commands, and all the commands are managed by this **API**.
+
+*  **Location**: */core/common/assets/js/components/commands.js*
+
+*  **Parent**: `elementorModules.Module`
+
+* **Available commands:** for further information about **all** the commands, please visit [`{$e.commands.getAll()}`](commands-methods/getall.md)**method**.
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply docs/core/common/assets/js/api/core/commands.md:12]: Not exactly at this line, its acts as a link.

File: docs/core/common/assets/js/api/core/hooks.md (hunk 15 above)
Comment (iNewLegend) [reply docs/core/common/assets/js/api/core/hooks.md]: Handled.

File: docs/core/common/assets/js/api/core/hooks.md (hunk 16 above)
Comment (iNewLegend) [reply docs/core/common/assets/js/api/core/hooks.md]: Handled.

File: docs/core/common/assets/js/api/core/components.md [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+*  **Methods**:
+
+	| Method                       | Parameters                     | Returns                                       | Description
+	|------------------------------|--------------------------------|-----------------------------------------------|------------------------------|
+	| `$e.components.getAll()`     |                                | `{array.<string>}`                            | Receive all components.
+	| `$e.components.register()`   | `{ComponentBase}` *component*  | `{ComponentBase}` *component*                 | Register a component.
+	| `$e.components.get()`        | `{String}` *id*                | `{ComponentBase}` *component*                 | Get component instance by id.
+	| `$e.components.getActive()`  |                                | `{Object.<ComponentBase>}` *activeComponents* | Get active components.
+	| `$e.components.activate()`   | `{String}` *namespace*         |                                               | Activate component.
+	| `$e.components.inactivate()` | `{String}` *namespace*         |                                               | Deactivate component.
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply docs/core/common/assets/js/api/core/components.md:22]: @matipojo, please help, what should we do?

File: docs/core/common/assets/js/api/core/components.md [hunk 21]
```diff
@@ -0,0 +1,116 @@
+# API - `$e.components`
+The new Components API (since 2.7.0), provides a simple and convenient way to bind all route and commands, and keyboard shortcuts that belong to a UI component – into one controller.
+The full list of components, including custom & 3rd routes, is available via: `$e.components.getAll();`
+Each component has it’s unique namespace, that all it’s command and routes are nested under it.
+
+On route to a component route, the component becomes “active” that mean that allows make sure that keyboard shortcuts of this component will work only if the component is active.
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply docs/core/common/assets/js/api/core/components.md:6]: Now is better?

File: docs/core/common/assets/js/api/core/commands.md [hunk 22]
```diff
@@ -0,0 +1,174 @@
+## API --  `$e.commands`
+The new Commands API (since 2.7.0), provides a simple and convenient way to run something the the editor, create a widget, as well as show a notice or undo changes, using JS commands.
+
+The full list of commands, including custom & 3rd commands, is available via: `$e.commands.getAll();`
+
+*  **Description**: `$e.commands` API is a manager of `$e.commands`, allow you to create custom **commands** that runs by `$e.run()`, each components have his own unique commands, and all the commands are managed by this **API**.
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply docs/core/common/assets/js/api/core/commands.md]: Now is better?

File: docs/core/common/assets/js/api/core/commands.md [hunk 23]
```diff
@@ -0,0 +1,174 @@
+## API --  `$e.commands`
+The new Commands API (since 2.7.0), provides a simple and convenient way to run something the the editor, create a widget, as well as show a notice or undo changes, using JS commands.
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply docs/core/common/assets/js/api/core/commands.md]: Handled.

File: docs/core/common/assets/js/api/core/components.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (matipojo) [reply docs/core/common/assets/js/api/core/components.md:22]: https://wikidiff.com/inactivate/deactivate

**Review by matipojo — DISMISSED** [review]

File: docs/core/common/assets/js/api/core/commands.md [hunk 24]
```diff
@@ -0,0 +1,174 @@
+## API --  `$e.commands`
+The new Commands API (since 2.7.0), provides a simple and convenient way to run something the editor, create a widget, as well as show a notice or undo changes, using JS commands.
```
Comment (kobizz) [inline docs/core/common/assets/js/api/core/commands.md:2]: > to run something the editor

Probably you mean "to run something *on/in* the editor

File: docs/core/common/assets/js/api/core/components.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (kobizz) [reply docs/core/common/assets/js/api/core/components.md]: Try something like "which allows binding of keyboard shortcuts..."

[the remark this answers is not in the document]
Comment (iNewLegend) [reply docs/core/common/assets/js/api/core/components.md]: handled.

File: docs/core/common/assets/js/api/core/commands.md (hunk 24 above)
Comment (iNewLegend) [reply docs/core/common/assets/js/api/core/commands.md]: Handled.

**Review by kobizz — APPROVED** [review]

**Review by matipojo — APPROVED** [review]

**Review by KingYes — APPROVED** [review]

**Pull request merged**